stan/math/prim/prob/wishart_cholesky_lpdf.hpp

```diff
@@ -0,0 +1,98 @@
+namespace math {
+
+/** \ingroup multivar_dists
+ * Return the natural logarithm of the unnormalized Wishart density of the
+ specified
+ * lower-triangular Cholesky factor variate, positive degrees of freedom, and
+ lower-triangular
+ * Cholesky factor of the scale matrix.
+ *
+ * The scale matrix, L_S, must be a lower Cholesky factor
+ * dimension, k, is implicit
+ * nu must be greater than k-1
+ *
+ * The change of variables from the input positive-definite matrix to
+ * the Cholesky factor is given in Theorem 2.1.9 in
+ * Muirhead, R. J. (2005).
+ * Aspects of Multivariate Statistical Theory. Wiley-Interscience.
+
+ * @tparam T_y Cholesky factor matrix
+ * @tparam T_dof scalar degrees of freedom
+ * @tparam T_scale Cholesky factor matrix
+ * @param L_Y lower triangular Cholesky factor of the inverse covariance matrix
+ * @param nu scalar degrees of freedom
+ * @param L_S lower triangular Cholesky factor of the scale matrix
+ * @return natural logarithm of the Wishart density at L_Y given nu and L_S
+ * @throw std::domain_error if nu is not greater than k-1
+ * @throw std::domain_error if L_S is not a valid Cholesky factor
+ */
+template <bool propto, typename T_y, typename T_dof, typename T_scale,
+          require_stan_scalar_t<T_dof>* = nullptr,
+          require_all_matrix_t<T_y, T_scale>* = nullptr>
+return_type_t<T_y, T_dof, T_scale> wishart_cholesky_lpdf(const T_y& L_Y,
+                                                         const T_dof& nu,
+                                                         const T_scale& L_S) {
+  using Eigen::Dynamic;
+  using Eigen::Lower;
+  using Eigen::Matrix;
+  using T_L_Y_ref = ref_type_t<T_y>;
+  using T_nu_ref = ref_type_t<T_dof>;
+  using T_L_S_ref = ref_type_t<T_scale>;
+  static const char* function = "wishart_cholesky_lpdf";
+  Eigen::Index k = L_Y.rows();
+  check_size_match(function, "Rows of random variable", L_Y.rows(),
+                   "columns of scale parameter", L_S.rows());
+
+  T_L_Y_ref L_Y_ref = L_Y;
+  T_nu_ref nu_ref = nu;
+  T_L_S_ref L_S_ref = L_S;
+
+  check_greater(function, "Degrees of freedom parameter", nu_ref, k - 1);
+  check_positive(function, "Cholesky Random variable", L_Y_ref.diagonal());
+  check_positive(function, "columns of Cholesky Random variable",
+                 L_Y_ref.cols());
+  check_positive(function, "Cholesky Scale matrix", L_S_ref.diagonal());
+  check_positive(function, "columns of Cholesky Scale matrix", L_S_ref.cols());
+
+  return_type_t<T_y, T_dof, T_scale> lp(0.0);
+
+  if (include_summand<propto, T_dof>::value) {
+    lp += k * LOG_TWO * (1 - 0.5 * nu_ref);
+  }
+
+  if (include_summand<propto, T_dof>::value) {
+    lp -= lmgamma(k, 0.5 * nu_ref);
+  }
+
+  if (include_summand<propto, T_dof, T_scale, T_y>::value) {
+    auto L_SinvL_Y = mdivide_left_tri<Eigen::Lower>(L_S_ref, L_Y_ref);
+    for (int i = 0; i < k; i++) {
+      lp -= 0.5 * dot_self(L_SinvL_Y.row(i).head(i + 1))
+            + nu_ref * log(L_S_ref.coeff(i, i))
```

Comment:
[optional]
Some of this can be made more efficient.

1. multiply by 0.5 after adding all the dot self's.  Otherwise, multiplication is a bit expensive originally and during autodiff.
2.  remove line 82 and include on outside as. ``lp -= nu_ref * L_S_ref.diag().log().sum();` or whatever the relevant Eigen calls are to do that.  It again pulls the multiplication out of the loop.
3. cache `nu_ref - 1` and reuse in 83.  This also saves auutodiff.  
4. replace `a * log(b)` with `multiply_log(a, b)`---it should have more efficient built-in autodiff.
5.  pull line 83 out and replace with `lp -= RowVectorXd::LinSpaced(1+nu_ref, 1+nu_ref + k) * L_Y_ref.diag().log()`. 

I may have slipped a sign.  I'd generally recommend always using `+=` in these compounds and just negating inside where necessary.  So rather than `lp -= a + b - c`, I'd recommend `lp += c - a - b;` to have less nested negation.

Reply:
On point 2.
One of the tests fails doing this because I'm passing an `unsigned int` for the dof. When I multiply this by the sum of the log diagonal it blows up. Turns out it's not doing the casting to a float. If I also multiply by `1.0` then it passes. Is there a better way to get this to work?

Reply:
On point 5.
Since it's a double `-` they cancel out and the correct replacement is 
```c++
// using stan::math functions
lp += stan::math::reverse(stan::math::linspaced_row_vector(k, nu_ref - k, nu_ref - 1)) * log(L_Y_ref.diagonal());

// using Eigen
lp += Eigen::RowVectorXd::LinSpaced(k, nu_ref - k, nu_ref - 1).reverse() * log(L_Y_ref.diagonal());
```
However, I cannot get this to work with the mix tests. Lots of stan-var issues. Do you have any suggestions?

Reply:
I did what I was able to. I can't get the `LinSpaced` to work with rev AD.

Reply:
Oddly it's just that `LinSpaced` function. It works if I create the row vector and populate it as below
```cpp
  if (include_summand<propto, T_dof, T_scale, T_y>::value) {
    auto L_SinvL_Y = mdivide_left_tri<Eigen::Lower>(L_S_ref, L_Y_ref);
    T_v dot_LSinvLY(0.0);
    Eigen::Matrix<T_v, 1, Eigen::Dynamic> linspaced_v(k);
    auto nu_minus_1 = nu_ref - 1;

    for (int i = 0; i < k; i++) {
      dot_LSinvLY += dot_self(L_SinvL_Y.row(i).head(i + 1));
      linspaced_v(i) = nu_minus_1 - i;
    }
    lp += -0.5 * dot_LSinvLY;
    lp += linspaced_v * log(L_Y_ref.diagonal());
    lp += -1. * nu_ref * sum(log(L_S_ref.diagonal()));
  }
```
Let me know if you'd prefer this in the PR.

Reply:
Thanks for working so hard on trying to get these to work.  

Point 2:  The standard way is to use `static_cast<double>(x)` rather than `x * 1.0`.  I wonder why it's not getting cast properly.  C++ should take care of any casting required for binary operators.

Point 5:  No worries if you can't make it work.  I'd think if this works, then a `.eval()` would also work (probably after `.reverse()`.

Reply:
Even that static cast is giving me the same errors. Maybe I'm doing something wrong. Here's what it errors
```cpp
./stan/math/prim/prob/wishart_cholesky_lpdf.hpp:76:12: error: cannot convert 'const stan::math::fvar<stan::math::fvar<double>>' to 'double' without a conversion operator
    lp += -static_cast<double>(nu_ref) * sum(log(L_S_ref.diagonal()));
```

Reply:
Oh, that's an `fvar<fvar<double>>`, so you can't cast to `double`.  Without more context, it's really hard to help debug here.  I can't figure out what multiplying by 1.0 is doing, because that will also give you an `fvar<fvar<double>>`.

Reply:
Ok, I'm going put that part back into the loop and hopefully can get this merged this week. I don't fully understand what you mean or what's happening. What I do understand is that the template type for `nu` is not fully capturing the complexity of what `nu` can be. Somehow I'd need to tell the compiler that `nu` can be a bunch of things, even an `fvar<fvar<double>>`. 

I'd be grateful if you'd take a look or if we could meet and take a look together so I have a better understanding of this type of template issue in the future.

Reply:
I got it to work!

Reply:
Always glad to supply a coding puzzle :-).   Advanced solutions would require someone else.